<launch>
  <!-- arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="map_file" default="$(find turtlebot3_navigation)/maps/map.yaml"/>
  <arg name="open_rviz" default="true"/>

  <!-- robot namespace -->
  <arg name="robot_namespace_01" default="turtlebot01"/>
  <arg name="robot_namespace_02" default="turtlebot02"/>
  <arg name="robot_namespace_03" default="turtlebot03"/>
    
  <!-- map server -->
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)" respawn="false">
    <param name="frame_id"    value="map"/>
  </node>

  <group ns = "$(arg robot_namespace_01)">

    <!-- amcl -->
    <node pkg="amcl" type="amcl" name="amcl">
      <rosparam file="$(find multi_turtlebot3_environment)/param/amcl_params.yaml" command="load" />
      <param name="initial_pose_x"   value="-2.0"/>
      <param name="initial_pose_y"   value="-0.5"/>

      <param name="odom_frame_id"    value="$(arg robot_namespace_01)/odom"/> 
      <param name="base_frame_id"    value="$(arg robot_namespace_01)/base_footprint"/>

      <!-- Use the single map server -->
      <remap from="static_map" to="/static_map"/>
    </node>

    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
      <!-- Set planner -->
      <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>

      <!-- Default configs form the TB3 repos -->
      <rosparam file="$(find multi_turtlebot3_environment)/param/costmap_common_params_$(arg model).yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find multi_turtlebot3_environment)/param/costmap_common_params_$(arg model).yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find multi_turtlebot3_environment)/param/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find multi_turtlebot3_environment)/param/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find multi_turtlebot3_environment)/param/move_base_params.yaml" command="load" />
      <rosparam file="$(find multi_turtlebot3_environment)/param/dwa_local_planner_params_$(arg model).yaml" command="load" />

      <!-- Set tf_prefix for frames explicity, overwriting defaults -->
      <!---->
      <param name="global_costmap/scan/sensor_frame"                value="$(arg robot_namespace_01)/base_scan"/>
      <param name="global_costmap/obstacle_layer/scan/sensor_frame" value="$(arg robot_namespace_01)/base_scan"/>
      <param name="global_costmap/global_frame"                     value="map"/>
      <param name="global_costmap/robot_base_frame"                 value="$(arg robot_namespace_01)/base_footprint"/>
      <param name="local_costmap/scan/sensor_frame"                 value="$(arg robot_namespace_01)/base_scan"/>
      <param name="local_costmap/obstacle_layer/scan/sensor_frame"  value="$(arg robot_namespace_01)/base_scan"/>
      <param name="local_costmap/global_frame"                      value="$(arg robot_namespace_01)/odom"/>
      <param name="local_costmap/robot_base_frame"                  value="$(arg robot_namespace_01)/base_footprint"/>
      
      <!-- Centralized map server -->
      <remap from="map" to="/map"/>
    </node>

  </group>

  <group ns = "$(arg robot_namespace_02)">

    <!-- amcl -->
    <node pkg="amcl" type="amcl" name="amcl">
      <param name="initial_pose_x"   value="-2.0"/>
      <param name="initial_pose_y"   value="0.5"/>

      <param name="odom_frame_id"    value="$(arg robot_namespace_02)/odom"/> 
      <param name="base_frame_id"    value="$(arg robot_namespace_02)/base_footprint"/>

      <!-- Use the single map server -->
      <remap from="static_map" to="/static_map"/>
    </node>

  </group>

  <group ns = "$(arg robot_namespace_03)">

    <!-- amcl -->
    <node pkg="amcl" type="amcl" name="amcl">
      <param name="initial_pose_x"   value="2.0"/>
      <param name="initial_pose_y"   value="-0.5"/>

      <param name="odom_frame_id"    value="$(arg robot_namespace_03)/odom"/> 
      <param name="base_frame_id"    value="$(arg robot_namespace_03)/base_footprint"/>

      <!-- Use the single map server -->
      <remap from="static_map" to="/static_map"/>
    </node>

  </group>

  <!-- rviz -->
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
          args="-d $(find multi_turtlebot3_environment)/rviz/multi_turtlebot3_navigation.rviz"/>
  </group>
</launch>
